src/scripts/hunt-filter/engine.ts

```diff
@@ -0,0 +1,129 @@
+
+    public validateUser(pre: User, post: User): boolean {
+        return this.userRules.every(r => {
+            let isValid = r.isValid(pre, post);
+            if (!isValid) {
+                this.logger.debug(`User object invalid due to rule: ${r.name}`);
+            }
+            return isValid;
+        });
+    }
+
+    public validateMessage(pre: IntakeMessage, post: IntakeMessage): boolean {
+        // TODO: Refactor? Location is being set to null in stage funcs so it'll fail later
+        if (pre.location === null || post.location === null) {
+            return false;
+        }
+
+        // Run rules. Build set of currently invalid properties
+        // { "location", "stage" }
+        const invalidProperties = new Set<(keyof IntakeMessage)>();
+        for (const rule of this.messageRules) {
+            const valid: boolean = rule.isValid(pre, post);
+            if (!valid) {
+                this.logger.debug(`Message possibly invalid due to rule: ${rule.name}`);
+                invalidProperties.add(rule.property);
+            }
+        }
+
+        // Don't have to run exemption rules if there are no violations
+        if (invalidProperties.size === 0) {
+            return true;
+        }
+
+        // Find exception objects that will run for given key
+        const relevant_exemptions_providers = this.messageExemptions.filter(e => invalidProperties.has(e.property));
+        this.logger.debug(`Got ${relevant_exemptions_providers.length} exemption providers for these invalid properties:`, ...invalidProperties.values());
+
+        // Do we need to filter them furthur? 
+        // Exception object could give specific key and value to filter on.
+        // Didn't pre-optimize but could be worth if there are many, many exemptions and they are slow (doubtful)
+        for (const exemption_provider of relevant_exemptions_providers) {
+            // Each exemption can give multiple keys that it accounts for.
+            // For example, the location and stage will change when catching realm ripper
+            // so that exemption provides [ "location", "stage" ]
+            const exemptions = exemption_provider.getExemptions(pre, post);
+            if (exemptions && exemptions.length > 0) {
+                this.logger.debug(`Generated exemptions from ${exemption_provider.name}`, { properties: exemptions });
+                exemptions.forEach(e => invalidProperties.delete(e));
+            }
+
+            if (invalidProperties.size == 0) {
+                this.logger.debug('Message was revalidated due to exemptions.');
+                return true;
+            }
+        }
+
+        this.logger.debug(`Message object invalid`, {
+            properties: { ...invalidProperties.values() },
+            messages: { pre, post }
+        })
+        
+        return false;
+    }
+
+    private initResponseRules() {
+        this.responseRules = [
+            new ResponseRules.ApiResponseBothRequireSuccess(),
```

Comment:
Is there a reason to require creating a rule and also remembering to create it here? Should the default export from the rule files be an array of all the rules instead?

Reply:
Good idea. I will see what I can do.

Reply:
I _can_ do this, however I think we lose the ability to inject dependencies via the constructor (maybe someone needs `ILogger` later). We can go back if needed though. Nothing lost.